### ISSUE TITLE
Modules: Testing code likely needs special powers

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/language#1698]
The in progress Module feature currently has the following section:

```
The rules for using types outside of their module are based on the capabilities the type explicitly provides:

It is a compile-time error to invoke a generative constructor of an abstract class, interface, or mixin outside of the module where the type is defined.

It is a compile-time error for a type to appear in an extends clause outside of the module where the type is defined unless the type is marked open.

It is a compile-time error for a type to appear in an implements clause outside of the module where the type is defined unless the type is an interface or is marked interface.

It is a compile-time error for a type to appear in a with clause outside of the module where the type is defined unless the type is a mixin or is marked mixin.
```

However, testing code within the same package most likely wants to violate these rules. And possibly even other modules within the same package that aren't just tests.

The most common case for tests would be Mocks or Fakes, which need to be able to `implement` all kinds of classes that exist in the same package, regardless of their stated rules.

## Friend modules?

One way to resolve this would be "friend modules". Friend modules would only be configurable from the originating module, maybe via some "friend <name>;"  directive or other similar mechanism. Any module listed as a friend would have the same capabilities as the current module.

**Comment by jakemac53** [comment]
Friend modules would also give you a way of providing an interface that is actually available to other packages, but only the ones you specifically mention in _your_ package. This allows you to have control over the implementations of the interface, and know exactly who is depending on the details of it, and in exactly what ways. You can push pre-emptive fixes to them or ensure they keep a tighter constraint on your package than other dependencies would, etc.

**Comment by lrhn** [comment]
Making test belong to the same "module" as the code it tests is something we already handle for packages.

If you declare your library's module like, fx:
```dart
module some.words;
```
and all libraries with the same module ID (package scoped) are in that module, then test files could belong to the module they're testing. They already belong to the package the module is in, so that's consistent.

**Comment by jakemac53** [comment]
I think it would be highly problematic for them to be a part of an existing module:
 
- This will create dependency cycles - the test will have extra dependencies which might create cycles (the module itself now has all those dependencies).
- A single test will want to mock many classes in a package (from different modules), meaning it would have to be a part of multiple modules, which doesn't really make sense imo. 
- This means the module isn't consistent depending on if you are in a testing context or a package context. There is no distinction here in the bazel/build_runner build systems, and it would be very problematic for them.

**Comment by munificent** [comment]
I agree with Jake. I think we don't want tests to be in the *same* module. The smaller your modules and the fewer the dependencies they have, the better. I think something friend modules could be a good solution, though I'm generally afraid of adding much complexity to this beyond the minimum needed. A few other ideas:

* We could loosen the class capability restrictions to only apply at the *package* level, instead of the module level. So all modules in the same package can do whatever they want with any class in any module in the same package.

* We could add modifiers like "interface for tests" that allow certain class capabilities *only* to other test modules in the same package.

* @leafpetersen suggested this morning that we could allow modules to nest and have some of the restrictions loosened for certain parent-child or sibling modules.

I'm somewhat hesitant about all of these, though. @leafpetersen pointed out that if we want these capability restrictions to affect static analysis then loosening them for tests could make that weird. You would conceivably get different static checking in tests than you would in other code. Or the static analysis would simply become unsound.

I *do* want to use these features for static checking. In particular, my hope is that a sealed family of subclasses can be used for algebraic-datatype style programming with pattern matching. In that case, I really really want exhaustiveness checking on those subclasses. This is particularly important if we want a pattern match *expression* since it needs to reliably have some case match so that it can yield a value. Exhaustive on match expressions over a family of subclasses is only sound if it really is a closed family of subclasses.

Also, I could see us possibly building on some of these restrictions in order to get better promotion on fields. (For example, a final field on a sealed non-interface class could be safe to promote.)

I think it's worth keeping in mind that there are two valid "solutions" to this problem:

1. Simply don't implement arbitrary random classes in your tests. It's not a foregone conclusion that every class in the world needs to be mocked. Our data from @bwilkerson indicates that only ~1% of classes are implemented inside test libraries. (That may not include generated code, though.) It may be fine to say that, yes, there are simply some classes that you can't mock in tests.
 
 2. Conversely, if you do want to implement it, just allow the class to be implemented. Open it up for your tests and let the rest of the world implement it too (with possibly a doc comment asking them not too). That is the state of the world today for *all* classes and it doesn't seem particularly untenable.

Given those, it might be sufficient to not do anything special for tests. Just have users not implement the classes they don't want implementable and let anyone implement the classes they do. That's not *ideal*. I am interested in being able to more gracefully support white box tests without exposing a module's innards to the world. But if we can't make that work somehow, I think we can get by without it.

**Comment by leafpetersen** [comment]
> I _do_ want to use these features for static checking. In particular, my hope is that a sealed family of subclasses can be used for algebraic-datatype style programming with pattern matching. In that case, I really really want exhaustiveness checking on those subclasses. This is particularly important if we want a pattern match _expression_ since it needs to reliably have some case match so that it can yield a value. Exhaustive on match expressions over a family of subclasses is only sound if it really is a closed family of subclasses.

I hesitate to suggest this because I think it adds a lot of complexity and I'm not really a fan of it, but for completeness... one could potentially consider defining two different modes of compilation: one in which exhaustive matches/promotions are *trusted* by the compiler, and one where the they are implicitly checked (via an implicitly added default case, or checks on the uses).  The latter mode is available only when compiling/running tests, but allows code in specific adjacent modules ("friend modules") to violate sealing.

**Comment by lrhn** [comment]
Dart originally had a "No modes!" creed. We've gone away from that, but every mode is a risk of introducing discrepancies between modes. A special mode just for *testing* risks that the code and behavior that you test, is not going to match the code that you actually deploy. I'd very much like to avoid that situation. The less special tests are, the better they *test*.

Even allowing tests to do things that non-tests cannot do, and have the occurrence of those things potentially affect global compilation strategies, runs some of  the same risks. (In that case, it's probably only compiler bugs that would cause a problem, not end-user code, since adding more code shouldn't affect compilation semantics, at most performance).

I worry about "friend modules". The need for that suggests to me that the abstraction is not at the right level, or doesn't have the correct affordances, if you need to be able to break it.
What if you could instead have nested modules, module hierarchies, where higher level modules depend on their nested sub-modules and can also ignore the nested modules' restrictions. 
It's still only inside the same package, but the *test* module could then contain all other modules of the package, and can override their restrictions.

I know it's all wrong wrt. compilation strategy, you can't see, when compiling the sub-module alone, whether a super-module breaks its abstractions. 
Friend modules seemingly have the correct ordering, you can see in the module which other modules might break its abstractions, but it still means you need to look at those to see whether they do, and those other modules definitely do depend on the original, so it's an implicit cycle.

I don't see a good solution where a test can break abstractions of a module that compilation of the module might want to depend on, but the test depends on the module, not the other way around. It *is* a cyclic compilation dependency if compilation of a module depends on a friend module's behavior, and the friend module depends on the original module.
Heck, even having the `lib/` module *know* about files in `/test` is questionable. The dependency direction is just wrong.

**Comment by jakemac53** [comment]
It seems to me that there are two major motivating factors for wanting to add access control modifiers to the language:

- Code organization & safety - having more confidence in what is a breaking change or not
- Compiler optimizations, field promotion, etc

The first would still be fine with friend modules, but I do fear we would break the latter use case in important ways, especially with modular compilation where the compiler can't actually see the friend modules.

Actually, I think the fact that modules currently allow other modules in the same package to violate the rules is also problematic in the same way?

**Comment by munificent** [comment]
> I hesitate to suggest this because I think it adds a lot of complexity and I'm not really a fan of it, but for completeness... one could potentially consider defining two different modes of compilation: one in which exhaustive matches/promotions are _trusted_ by the compiler, and one where the they are implicitly checked (via an implicitly added default case, or checks on the uses).

This isn't a totally crazy idea to me. It's basically an test-only assertion, and we already have the distinction that assertions run in tests but usually not in production code.

> The less special tests are, the better they _test_.

Agreed, totally. If we do something special for tests or test-like modules I want it to be minimally invasive and/or part of some larger module related feature not specific to tests. ("Friend modules" might have tests as a main use case, but they could conceivably be used to solve other problems too.)

> Actually, I think the fact that modules currently allow other modules in the same package to violate the rules is also problematic in the same way?

The current proposed behavior is that restrictions are only relaxed for libraries in the same module, not other modules in the same package. Basically, within a single unit of compilation we can relax the rules because that is the "whole world" from the perspective of that compilation.
